File: frontend/src/models/data/SinDivider.ts
import DataDivider from "./DataDivider";

export default class SinDivider extends DataDivider {
    wavelength: number;
    amplitude: number;
    constructor() {
        super();
        this.wavelength = 1;
        this.amplitude = 1;
    }

    evaluate(x: number, y: number): boolean {
        const theta = this.deg * Math.PI / 180;
        const sinT = Math.sin(theta);
        const cosT = Math.cos(theta);
        return cosT * (y + this.shiftY) - sinT * (x - this.shiftX) < Math.sin((sinT * (y + this.shiftY) + cosT *(x-this.shiftX)*this.wavelength))*this.amplitude;
    }
}

File: backend/Models/SinDivider.cs
namespace Graphic_Neural_Network.backend.Models
{
    using System;

    public class SinDivider : DataDivider
    {
        public override string Type => "sindivider";
        public double Wavelength { get; set; }
        public double Amplitude { get; set; }

        public SinDivider()
        {
            Wavelength = 1;
            Amplitude = 1;
        }

        public override bool Evaluate(double x, double y)
        {
            double theta = Deg * Math.PI / 180;
            double sinT = Math.Sin(theta);
            double cosT = Math.Cos(theta);

            double shiftedY = y + ShiftY;
            double shiftedX = x - ShiftX;

            double projection = sinT * shiftedY + cosT * shiftedX;
            return cosT * shiftedY - sinT * shiftedX < Math.Sin(projection * Wavelength) * Amplitude;
        }
    }

}
